Update the STM32 USB driver to the new API

c1728p9 opened the pull request:
Take the code from mbed-os\features\unsupported\USBDevice\targets\TARGET_STM as a starting point and use it to fill in the USBPhy template for STM32 devices.

c1728p9 committed:
Update the STM32 USB driver to the new API

Take the code from
mbed-os\features\unsupported\USBDevice\targets\TARGET_STM
as a starting point and use it to fill in the USBPhy template for STM32
devices.

c1728p9 commented on usb/device/targets/TARGET_STM/USBPhy_STM32.cpp:
```diff
@@ -0,0 +1,522 @@
+    _read_buf[EP_TO_LOG(endpoint)] = data;
+    _read_size[EP_TO_LOG(endpoint)] = size;
+
+    uint8_t* pBuf = (uint8_t *)pBufRx; //TODO - this buffer shouldn't be shared for multiple endpoints
```
The same buffer is used for all endpoints which will corrupt data. Any way this can be done with the buffer passed in rather than a dedicated RX buffer?

jamike replied:
The dummy buffer pBuffRx is there mainly because, in previous API ,the read function does not have buffer in parameter. Only the read result function was having a buffer. Now with the new API , it is the opposite. So the buffer (data) can be used instead of the dummy buffer pBufRx  that was used for any request.

c1728p9 replied:
@jamike I have removed these buffers and am now passing data directly into the ST hal directly. There aren't any alignment requirements on the data is there?

c1728p9 commented on usb/device/targets/TARGET_STM/USBPhy_STM32.cpp:
```diff
@@ -0,0 +1,522 @@
+
+void USBPhyHw::endpoint_abort(usb_ep_t endpoint)
+{
+    // TODO - stop the current transfer on this endpoint and don't call the IN or OUT callback
```
I didn't see a function to abort the current transfer in the ST cube library. Does it have this functionality, and if not, is there a way to add it? This will be required if the buffer passed into endpoint_read can be used directly.

jamike replied:
Agree so the buffer read must be removed from Hal internal descriptor to avoid the copy on a memory area no more allocated for usb buffer reception , and the current transaction that has been programmed but not yet performed must be stopped, to allow the queuing of a new read request on same endpoint.  There is no function implemented this abort, a  new function HAL abort should disable the endpoint.  The function For F2,  will be something like the code in following commit https://github.com/jamike/mbed/commit/005e6b379acd043cb3e1699531c0045c389a2928
, this needs to be tested.

c1728p9 replied:
@jamike after fixes the return code check in endpoint_abort I was able to use your commit and pass testing. Endpoint abort tests aren't complete though so it wasn't exhaustively tested.

Would it be possible to get an endpoint abort function for all ST devices in a future release of the ST Cube library?

c1728p9 commented on usb/device/targets/TARGET_STM/USBPhy_STM32.cpp:
```diff
@@ -0,0 +1,522 @@
+    instance->events->start_process();
+}
+
+//TODO - remove this hack which allows HAL_Delay to work when interrupts are disabled
```
The ST cube library implemented HAL_Delay by polling a counter that was incremented by a ticker. Since USB is protected by a critical section this interrupt never fired causing a hang.

Furthermore since this is in a critical section any substantial delay could have a serious impact on interrupt latency. Any ideas on how to prevent blocking?

jamike replied:
To solve this , the HAL_Delay used under critical section ( if some , I searched  HAL_Delay, and I did not find a HAL_Delay call from USB interrupt handler)   must  not rely on tick interruption, it should read a counter value till a counter timeout value is reached.

c1728p9 replied:
The whole USBDevice class is protected with a critical section, so none of the calls can block. There were 3 places in the code with delays. Once I removed these then everything worked normally:
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L188
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L910
https://github.com/ARMmbed/mbed-os/blob/e43d21d4ef6025e94e4567961cb87ed2f0474a26/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L923

Are these delays required by the USB peripheral itself, or just to make things easier for the user? The bottom two enable / disable the USB pullup, so I would imagine this delay is not required by the peripheral, but instead required for disconnect / connect to be properly detected on USB.

Do you have any recommendations on how to proceed with these delays @jamike? Here are some possible paths forward:
- Add non-blocking versions of these same functions to the ST library
- Remove these delays from ST library code
- Lazily delay - only delay in the second of two calls for the time remaining. If the caller delays externally then no blocking is needed.

bcostm replied:
I have also thinking to overwrite the HAL_Delay function (as it is declared weak) and instead call a function that will only read us_ticker_read(). Similar to PR #6538 but this wait loop function has to be declared in a place that we can use it for all STM devices or even better at mbed HAL level. But I don't know if it will solve your issue?

c1728p9 replied:
Hi @bcostm, that would solve my immediate issue of a lockup. Unfortunately, it would violate timing requirements, as this code is running with interrupts disabled.

c1728p9 reviewed (commented):
@jeromecoutant @LMESTM this should be a good starting point for USB support. It passes all the tests in their current form. There are 3 TODOs that still need to be resolved (this will cause future tests to fail). I left comments on this PR about these three remaining items. Please have a look and let me know if you have any thoughts.

c1728p9 committed:
Remove ep0_write from set_address

The USBDevice stack automatically sends control endpoint status
so this does not and should not be done from the call to set_address.

jamike committed:
Create HAL_PCD_EP_Abort

jeromecoutant commented on usb/device/targets/TARGET_STM/USBPhy_STM32.cpp:
```diff
@@ -0,0 +1,521 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
We should add something like
#if defined(DEVICE_USBDEVICE) && DEVICE_USBDEVICE

c1728p9 replied:
Done

c1728p9 committed:
Remove unnecessary USB receive buffer

Remove receive buffer from USBPhy_STM32 and instead passed buffers in
directly.

c1728p9 committed:
Fix return code check in endpoint abort

Assert that HAL_PCD_EP_Abort returns HAL_OK rather that something else.

c1728p9 committed:
Only build in USBPhy_STM32 for supported devices

Define USBSTM_HAL_UNSUPPORTED if DEVICE_USBDEVICE is undefined or
defined to 0.

c1728p9 commented on targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c:
```diff
@@ -745,6 +745,61 @@ HAL_StatusTypeDef USB_EP0StartXfer(USB_OTG_GlobalTypeDef *USBx , USB_OTG_EPTypeD
+    {
+      volatile uint32_t loop=0;
+      USBx_INEP(ep->num)->DIEPCTL |= (USB_OTG_DIEPCTL_EPDIS);
+      while(((USBx_INEP(ep->num)->DIEPCTL) & USB_OTG_DIEPCTL_EPENA) ==  USB_OTG_DIEPCTL_EPENA)
```
@jamike does this interrupt register need to be cleared before looping on it?

0xc0170 commented:
@jeromecoutant @jamike Are there any changes needed here?

c1728p9 commented:
There are still two points that need to be addressed before this can move forward:
1. Need ST support with the missing endpoint abort functionality in ST cube library
1. Need ST support addressing the three separate delays in the low level USB code in the ST cube library

CC @jeromecoutant @jamike

bcostm commented:
> 1. Need ST support with the missing endpoint abort functionality in ST cube library

I have requested it. ST_INTERNAL_REF 46145

> 2. Need ST support addressing the three separate delays in the low level USB code in the ST cube library

Just a thought. Can we use instead RTOS `Thread::wait()` function ?

jeromecoutant commented:
> I have requested it. ST_INTERNAL_REF 46145

but this can take a very long time... 
So how can we bypass this abort functionality ?

c1728p9 commented:
> So how can we bypass this abort functionality ?

The alternative is to use static read and write buffers for each endpoint, rather than the than user provided buffers. That way if a user calls abort and then frees its buffer the freed memory won't be accessed when the transfer completes. This has the downside of adding an extra copy and extra memory usage for the static buffers.

Static endpoint buffers is what the mbed 2 version used.  For example a transfer started by `endpointRead` stores data in the internal buffer and once complete a user supplied buffer is filled with the result in `endpointReadResult`.

> Can we use instead RTOS Thread::wait() function ?

Nope, it is inside a critical section so `Thread::wait()` can't be used either.

cmonr commented:
@bcostm @jeromecoutant @jamike Any other comments, or can this PR progress?

bcostm commented:
For point 1 (Endpoint abort functionality): we have no short term solution in the HAL driver.
For point 2 (delays in ST HAL USB file): there is a workaround proposed by @c1728p9 (redefining the HAL_GetTick function + other patch) or second solution to remove all those delays. But we are not sure if this will be a problem or not... What we miss is additional extensive USB tests that will assure us that removing those delays is OK.
But even if point 2 is solved, we'll still have point 1 open...

c1728p9 commented:
I rebased this branch on top of feature-hal-spec-usb-device

cmonr commented:
/morph build

cmonr commented:
@bcostm What does point 1 mean in reference to this PR?

bcostm commented:
Well, the endpoint abort is something that @c1728p9 required to move forward with this PR.

We have the following possibilities:
1) Add it in ST HAL. We have requested it to our HAL team but it will not be added soon.
2) Do it ourself in mbed. @jamike started something here https://github.com/jamike/mbed/commit/005e6b379acd043cb3e1699531c0045c389a2928 but it is not completed as I understood.  But @jamike cannot work for the moment on it.
3) Use static endpoint buffers as @c1728p9 proposed like it was in mbed 2.

I think the solution 3) is the most appropriate to go faster. What do you think @c1728p9 ?

0xc0170 commented:
@c1728p9 @ARMmbed/team-st-mcd Based on the suggestions, this PR is not  yet ready? Or good to go and comments will be addressed via separate PR?

bcostm commented:
For what I understood this PR is not ready to be merged until the possibility to abort an endpoint transmission is possible. But I am maybe wrong...

cmonr commented:
@jeromecoutant @jamike @c1728p9 ?

jeromecoutant commented:
Let's merge!
